### PR TITLE
feat: edit the current command line for bash, fish, and zsh

### LINKED ISSUE #245: feat: make the current cmdline editable in place
Sometimes I start typing a long complicated command and then I get annoyed with the lack of vim keybindings. My shell (fish) has a keybinding to open the current commandline in `$EDITOR` but then I lose sight of the scrollback, so its suboptimal.

`kitty-scrollback.nvim` has the potential to be useful here. Currently adding to the current commandline is easy, but there is no easy way to select, edit or replace the existing commandline.

### PR DESCRIPTION
**Pull request by mikesmithgh**
closes #245

TODO:
- [x] documentation
- [x] tests
- [x] demo

### CODE REVIEW

File: lua/kitty-scrollback/windows.lua [hunk 1]
```diff
@@ -71,8 +71,15 @@ end
 M.open_paste_window = function(start_insert)
   vim.cmd.stopinsert()
 
+  if vim.env.KITTY_SCROLLBACK_NVIM_EDIT_INPUT then
```
Comment (mikesmithgh) [inline lua/kitty-scrollback/windows.lua:74]: consider adding a mode env var that can be used in the future for other special cases

File: scripts/edit_command_line.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+for input_file; do true; done
+
+# after exiting this script and before it has been read by kitty-scrollback.nvim
+# the contents of the original input_file may be altered
+# avoid this by copying the input_file to a new file that will be referenced
+ksb_input_dir=$(mktemp -d)
+ksb_input_file="$ksb_input_dir/input.ksb_editcommand"
+cp "$input_file" "$ksb_input_file"
+
+kitty @ kitten /Users/mike/gitrepos/kitty-scrollback.nvim/python/kitty_scrollback_nvim.py \
```
Comment (IndianBoy42) [inline scripts/edit_command_line.sh:19]: Hardcoded your path here

Comment (IndianBoy42) [reply scripts/edit_command_line.sh:19]: It works after patching it manually though

**Comment by IndianBoy42** [comment]
related feature we might want is being able to scroll the scrollback buffer, or more generally a way to go to the scrollback buffer and keep the prompt window open at the same time, and then go back to the prompt window.

**Comment by IndianBoy42** [comment]
oh I just noticed the prompt window buffer filetype is bash, in this case we can use the correct filetype based on the shell. i think in general we can figure out what shell is running using `kitten @ ls` right?

File: scripts/edit_command_line.sh (hunk 2 above)
Comment (mikesmithgh) [reply scripts/edit_command_line.sh]: thanks for pointing this out! I just removed the hardcoded path. Could you try with latest?

**Comment by mikesmithgh** [comment]
> oh I just noticed the prompt window buffer filetype is bash, in this case we can use the correct filetype based on the shell. i think in general we can figure out what shell is running using `kitten @ ls` right?

good catch. so it pulls it from `vim.o.shell` https://github.com/mikesmithgh/kitty-scrollback.nvim/blob/main/lua/kitty-scrollback/windows.lua#L104

I think that may just be what your `SHELL` environment variable is set to? What results do you get for `echo $SHELL`.

There is an option to override this in kitty-scrollback.nvim `opts.paste_window.filetype`. But, I agree this looks like a bug even without this functionality. If I am in bash and then go to fish and open kitty-scrollback.nvim it has the wrong shell. I'll open an issue for this.

**Comment by mikesmithgh** [comment]
> related feature we might want is being able to scroll the scrollback buffer, or more generally a way to go to the scrollback buffer and keep the prompt window open at the same time, and then go back to the prompt window.

interesting, yeah I'd have to think about this. 

I suppose I could have the ability to not autoclose the floating paste window or an option to open a regular window instead of a floating window. Then you can just navigate back and forth.

**Comment by mikesmithgh** [comment]
@IndianBoy42 thanks for the feedback! I like the ideas

**Comment by mikesmithgh** [comment]
> > oh I just noticed the prompt window buffer filetype is bash, in this case we can use the correct filetype based on the shell. i think in general we can figure out what shell is running using `kitten @ ls` right?
> 
> good catch. so it pulls it from `vim.o.shell` https://github.com/mikesmithgh/kitty-scrollback.nvim/blob/main/lua/kitty-scrollback/windows.lua#L104
> 
> I think that may just be what your `SHELL` environment variable is set to? What results do you get for `echo $SHELL`.
> 
> There is an option to override this in kitty-scrollback.nvim `opts.paste_window.filetype`. But, I agree this looks like a bug even without this functionality. If I am in bash and then go to fish and open kitty-scrollback.nvim it has the wrong shell. I opened the issue #254 for this.

@IndianBoy42 FYI I just pushed a fix for this. It should pull the filetype from the shell that is configured in Kitty now.

**Comment by IndianBoy42** [comment]
> > oh I just noticed the prompt window buffer filetype is bash, in this case we can use the correct filetype based on the shell. i think in general we can figure out what shell is running using `kitten @ ls` right?
> 
> good catch. so it pulls it from `vim.o.shell` https://github.com/mikesmithgh/kitty-scrollback.nvim/blob/main/lua/kitty-scrollback/windows.lua#L104
> 
> I think that may just be what your `SHELL` environment variable is set to? What results do you get for `echo $SHELL`.

Oh I just remembered about this, I manually `set shell=bash` because sometimes plugins assume that `shell` is posix compatible, or bash compatible and try to use it to run scripts. So if `set shell=fish` they error out. Using `vim.env.SHELL` or kitty's shell detection makes sense. and it works :+1:

File: scripts/edit_command_line.sh (hunk 2 above)
Comment (IndianBoy42) [reply scripts/edit_command_line.sh]: works :+1:

**Comment by mikesmithgh** [comment]
`zsh 5.9` on my local is passing, `zsh 5.8` on CI is failing

**Comment by mikesmithgh** [comment]
Hooray tests are passing. 

TODO: update runners outside of this PR, test debug still works, update docs/generate command for require zsh 5.9

**Comment by mikesmithgh** [comment]
PR for github runner updates: https://github.com/mikesmithgh/kitty-scrollback.nvim/pull/288

Tested debugging github action with latest runner version 👍 

Screenshot 2024-12-18 at 7 50 00 AM

File: lua/kitty-scrollback/windows.lua (hunk 1 above)
Comment (mikesmithgh) [reply lua/kitty-scrollback/windows.lua]: will do this in the future if there is ever a case for this

**Pull request merged**